### PR TITLE
OS:13383062 Don't hoist instructions out of a loop if an argument isn't primitive in the landing zone

### PR DESCRIPTION
**Pull request by ricobbe**
Avoids a problem, found by ExprGen, where we have an instruction in the loop body (loop condition, actually) with an argument that's tagged as a float.  We're hoisting it out of the loop, despite the fact that the float tag is only valid because of an earlier instr in the loop condition, which bails if the arg isn't actually a float.  As a result, in this particular test case, the hoisted instruction ends up calling the argument's valueOf method too many times, and so its side effects happen too much.

### CODE REVIEW

**Comment by ricobbe** [comment]
@dotnet-bot test Tab Check please
@dotnet-bot test Ubuntu shared_ubuntu_linux_release please
@dotnet-bot test Ubuntu shared_ubuntu_linux_test
@dotnet-bot test Ubuntu static_ubuntu_linux_debug please
@dotnet-bot test Ubuntu static_ubuntu_linux_test
@dotnot-bot test Windows ci_disablejit_x64_debug

File: test/Bugs/rlexe.xml [hunk 1]
```diff
@@ -373,4 +373,11 @@
       <files>valueInfoLossBug.js</files>
     </default>
   </test>
+  <test>
+    <default>
+      <files>bug13383062.js</files>
+      <compile-flags>-maxinterpretcount:2 -off:simplejit -off:lossyinttypespec -bgjit- -off:redeferral</compile-flags>
```
Comment (rajatd) [inline test/Bugs/rlexe.xml]: >redeferral [](start = 92, length = 10)

this switch isn't really required for your test case. I had just used it to rule out redeferral

File: test/Bugs/bug13383062.js [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+        print('loop 2');
+        protoObj1.method1((class7.func56 = {
+            valueOf: function () {
+                WScript.Echo("class7.func56.valueOf");
+            }
+        }, class7.func56 = arrObj0));
+    }
+}
+test0();
+// command: D:\ReducerArena\chakrafs\fs\Builds\ChakraFull\unreleased\rs3\1708\00015.24208_170818.1733_michhol_4adec5c86c80a442c3cd8c1fcb25e51243115e8c\bin\x86_debug\20170818175152\jshost.exe -PrintSystemException  -es6generators -es7asyncawait  -maxinterpretcount:1 -maxsimplejitruncount:1 -MinMemOpCount:0 -werexceptionsupport  -force:atom -force:fixdataprops -force:rejit -off:cachedScope -off:stackargopt -off:BoundCheckHoist -off:trackintusage -off:ArrayCheckHoist -force:ScriptFunctionWithInlineCache -off:DelayCapture -oopjit- -ForceArrayBTree -off:fefixedmethods -off:aggressiveinttypespec -off:LoopCountBasedBoundCheckHoist -off:lossyinttypespec -off:checkthis -off:ParallelParse -ValidateInlineStack -off:ArrayLengthHoist step1267.js
```
Comment (rajatd) [inline test/Bugs/bug13383062.js]: nit: could remove the comments at the end to keep the test clean

File: lib/Backend/GlobOpt.cpp [hunk 3]
```diff
@@ -16142,8 +16142,17 @@ GlobOpt::OptIsInvariant(Sym *sym, BasicBlock *block, Loop *loop, Value *srcVal,
...
+    loopHeadVal = loop->landingPad->globOptData.FindValue(sym);
+    if (loopHeadVal == NULL || loopHeadVal->GetValueNumber() != srcVal->GetValueNumber())
+    {
+        return false;
+    }
+
+    // Can't hoist non-primitives, unless we have safeguards against valueof/tostring.  Additionally, we need to consider
+    // the value annotations on the source *before* the loop: if we hoist this instruction outside the loop, we can't
+    // necessarily rely on type annotations added (and enforced) earlier in the loop's body.
+    if (!allowNonPrimitives && !loopHeadVal->GetValueInfo()->IsPrimitive() && !loop->landingPad->globOptData.IsTypeSpecialized(sym))
```
Comment (rajatd) [inline lib/Backend/GlobOpt.cpp]: >loopHeadVal->GetValueInfo()->IsPrimitive() && !loop->landingPad->globOptData.IsTypeSpecialized [](start = 32, length = 94)

@louislaf, I wonder if these two checks check the same thing. Should an int/float type specialized sym have a definitely primitive value info?

File: test/Bugs/bug13383062.js (hunk 2 above)
Comment (ricobbe) [inline test/Bugs/bug13383062.js]: Fixed in commit 9ee182c

File: test/Bugs/rlexe.xml (hunk 1 above)
Comment (ricobbe) [inline test/Bugs/rlexe.xml]: Fixed in commit 64baf42

**Comment by ricobbe** [comment]
@dotnet-bot test Windows x86_debug

**Comment by mmitche** [comment]
@dotnet-bot test Windows x86_debug (machine was apparently an A0)

**Comment by ricobbe** [comment]
@dotnet-bot test Windows x64_debug

File: lib/Backend/GlobOpt.cpp (hunk 3 above)
Comment (LouisLaf) [reply lib/Backend/GlobOpt.cpp]: Yes, you are right.  The IsTypeSpecialized check is probably unnecessary and overly conservative.
Richard, can you try keeping only the IsPrimitive check to see what happens?


---
In reply to: [135135426](https://github.com/Microsoft/ChakraCore/pull/3581#discussion_r135135426) [](ancestors = 135135426)

Comment (ricobbe) [reply lib/Backend/GlobOpt.cpp]: Done; VSO builds both completed with no errors.

**Comment by ricobbe** [comment]
Unit tests, exprgen, and (local) perf tests look good after removing extraneous condition.  IB run still in progress.

**Comment by ricobbe** [comment]
Unit tests and local perf tests look good after restoring not-so-superfluous condition; ExprGen and IB still running.

**Comment by ricobbe** [comment]
Withdrawn in favor of PR #3624, q.v.

**Pull request closed without merging**